contrib/drivers/filesharedrivers/nfs/nfs.go

```diff
@@ -163,28 +165,59 @@ func (d *Driver) CreateFileShare(opt *pb.CreateFileShareOpts) (*model.FileShareS
 		return nil, err
 	}
 
-	if err := d.cli.CreateVolume(name, vg, opt.GetSize()); err != nil {
-		return nil, err
-	}
-	// remove created volume if got error
-	defer func() {
-		// using return value as the error flag
-		if fshare == nil {
-			if err := d.cli.Delete(name, vg); err != nil {
-				log.Error("failed to remove volume fileshare:", err)
-			}
+	if opt.SnapshotId != "" {
+		// User requested for creating fileshare using existing snapshot
+		//get fileshare name
+		var existingFsName = opt.GetMetadata()[KFileshareName]
+
+		// get existingfileshare snap logical path
+		//get volume group
+		var vg = opt.GetPoolName()
+		// create a fileshare device path
+		var lvPathForSnap = path.Join("/dev", vg, opt.SnapshotName)
+		// create a existing fileshare device path
+		var lvPathExistingPath = path.Join("/dev", vg, existingFsName)
+		// get directory where fileshare mounted
+		var olddirName = path.Join(MountPath, existingFsName)
+		// umount the volume to directory
+		if err := d.cli.UnMount(olddirName); err != nil {
+			log.Error("failed to mount a directory:", err)
+			return nil, err
 		}
-	}()
 
-	// Crete fileshare on this path
-	if err := d.cli.CreateFileShare(lvPath); err != nil {
-		log.Error("failed to create filesystem logic volume:", err)
-		return nil, err
-	}
-	// mount the volume to directory
-	if err := d.cli.Mount(lvPath, dirName); err != nil {
-		log.Error("failed to mount a directory:", err)
-		return nil, err
+		if err := d.cli.CreateFileShareFromSnapshot(lvPathForSnap); err != nil {
+			log.Error("failed to create filesystem from given snapshot:", err)
+			return nil, err
+		}
+		// mount the volume to directory
+		if err := d.cli.Mount(lvPathExistingPath, dirName); err != nil {
+			log.Error("failed to mount a directory:", err)
+			return nil, err
+		}
+	} else {
+		if err := d.cli.CreateVolume(name, vg, opt.GetSize()); err != nil {
+			return nil, err
+		}
+		// remove created volume if got error
+		defer func() {
+			// using return value as the error flag
+			if fshare == nil {
+				if err := d.cli.Delete(name, vg); err != nil {
+					log.Error("failed to remove volume fileshare:", err)
```

Comment:
Can it be a case that the volume was busy (some residual task)  and the deletion was ignored? Will having a retry mechanism help? Else, volume will be lying stale and add space?

Reply:
Actually this delete volume is not for healthy volume. If some volume creation got failed previously, then it will be deleted(clean up type of things). So this volume will not be busy in anycase because it's not in use.

Reply:
Ok

Reply:
If the creation failed, how the delete will be successful? Should we not have a cleanup call instead?